source/adios2/CMakeLists.txt

```diff
@@ -105,10 +104,24 @@ add_library(adios2_core
 set_property(TARGET adios2_core PROPERTY EXPORT_NAME core)
 set_property(TARGET adios2_core PROPERTY OUTPUT_NAME adios2${ADIOS2_LIBRARY_SUFFIX}_core)
 
+set(maybe_adios2_core_cuda)
 if(ADIOS2_HAVE_CUDA)
   enable_language(CUDA)
-  target_link_libraries(adios2_core PRIVATE CUDA::cudart)
-  set_target_properties(adios2_core PROPERTIES CUDA_SEPARABLE_COMPILATION ON)
+
+  add_library(adios2_core_cuda helper/adiosCUDA.cu)
+  set_target_properties(adios2_core_cuda PROPERTIES
+    CUDA_SEPARABLE_COMPILATION ON
+    CUDA_STANDARD 14
+    CUDA_STANDARD_REQUIRED ON
+    CUDA_VISIBILITY_PRESET hidden
+    VISIBILITY_INLINES_HIDDEN ON
+    INCLUDE_DIRECTORIES "$<BUILD_INTERFACE:${ADIOS2_SOURCE_DIR}/source>;$<BUILD_INTERFACE:${ADIOS2_BINARY_DIR}/source>"
+    EXPORT_NAME core_cuda
+    OUTPUT_NAME adios2${ADIOS2_LIBRARY_SUFFIX}_core_cuda
+    )
+
+  target_link_libraries(adios2_core PRIVATE adios2_core_cuda CUDA::cudart CUDA::cuda_driver)
+  set(maybe_adios2_core_cuda adios2_core_cuda)
 endif()
 
```

Comment:
So which line actually separated the linked symbols? Just curious.

Reply:
Thanks! Is this the only thing needed for separating symbols? I can see that as ADIOS2 grows, we will have this problem more and more frequently, for example, Engine A and Engine B both use Library C, but they require different versions.

Reply:
Yeah it is a somewhat obscure feature, once we make every symbol hidden we can select the ones that we want to export by prepending `__attribute__((visibility("hidden"))`, we can explore how this feature could benefit ADIOS2 in other modules in terms of binary size/building speed and removing UB